api/ghactions/notify.go

```diff
@@ -0,0 +1,200 @@
+	vars := mux.Vars(r)
+	rawRunID := vars["run_id"]
+	var runID int64
+	var err error
+	if runID, err = strconv.ParseInt(rawRunID, 0, 0); err != nil {
+		http.Error(w, fmt.Sprintf("Invalid run id: %s", rawRunID), http.StatusBadRequest)
+
+		return
+	}
+
+	owner := vars["owner"]
+	repo := vars["repo"]
+
+	if owner != shared.WPTRepoOwner || repo != shared.WPTRepoName {
+		http.Error(w, fmt.Sprintf("Invalid repo: %s/%s", owner, repo), http.StatusBadRequest)
+
+		return
+	}
+
+	artifactName := vars["artifact_name"]
+	artifactNameGlob, err := glob.Compile(artifactName)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Invalid artifact name: %s", artifactName), http.StatusBadRequest)
+
+		return
+	}
+
+	ctx := r.Context()
+	aeAPI := shared.NewAppEngineAPI(ctx)
+	log := shared.GetLogger(ctx)
+
+	ghClient, err := aeAPI.GetGitHubClient()
+	if err != nil {
+		log.Errorf("Failed to get GitHub client: %s", err.Error())
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+
+		return
+	}
+
+	processed, err := processBuild(
+		ctx,
+		aeAPI,
+		ghClient,
+		owner,
+		repo,
+		runID,
+		artifactNameGlob,
+	)
+
+	if err != nil {
+		log.Errorf("%v", err.Error())
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+
+		return
+	}
+
+	if processed {
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintln(w, "GitHub Actions workflow run artifacts retrieved successfully")
+	} else {
+		w.WriteHeader(http.StatusNoContent)
+		fmt.Fprintln(w, "Notification of workflow run artifacts was ignored")
+	}
+}
+
+func processBuild(
+	ctx context.Context,
+	aeAPI shared.AppEngineAPI,
+	ghClient *github.Client,
+	owner string,
+	repo string,
+	runID int64,
+	artifactNameGlob glob.Glob,
+) (bool, error) {
+	log := shared.GetLogger(ctx)
+
+	workflowRun, _, err := ghClient.Actions.GetWorkflowRunByID(ctx, owner, repo, runID)
+	if err != nil {
+		return false, err
+	}
+
+	headSha := workflowRun.HeadSHA
+
+	opts := &github.ListOptions{PerPage: 100}
+
+	uploadedAny := false
+	errors := make(chan error)
+
+	for {
+		artifacts, resp, err := ghClient.Actions.ListWorkflowRunArtifacts(ctx, owner, repo, runID, opts)
+
+		if err != nil {
+			return false, err
+		}
+
+		for _, artifact := range artifacts.Artifacts {
+
+			if !artifactNameGlob.Match(*artifact.Name) {
+				log.Infof("Skipping artifact %s", artifact.Name)
+
+				continue
+			}
+
+			log.Infof("Uploading %s for %s/%s run %v...", artifact.Name, owner, repo, runID)
+
+			labels := chooseLabels(workflowRun, *artifact.Name, owner, repo)
+
+			uploader, err := aeAPI.GetUploader(uploaderName)
+			if err != nil {
+				return false, fmt.Errorf("Failed to get uploader creds from Datastore: %w", err)
+			}
+
+			uploadClient := uc.NewClient(aeAPI)
+			err = uploadClient.CreateRun(
+				*headSha,
+				uploader.Username,
+				uploader.Password,
+				// Uploading a ZIP with both results and screenshots is special-cased by the receiver.
+				[]string{*artifact.ArchiveDownloadURL},
```

Comment:
I think this may actually be wrong. Azure Pipelines follows the older GitHub Actions behaviour of merging everything, whereas with GitHub Actions we end up with one ZIP per job (and we use N jobs to parallelise test execution).

Reply:
Yeah, this is definitely wrong and needs fixed. We need to be making a single call here per Workflow Run, not per Job.